include/charmlite/core/common.hpp

```diff
@@ -23,6 +23,35 @@ void CmiInitMemAffinity(char** argv);
 
 namespace cmk {
 
+    // Helper singleton class
+    template <typename T>
+    class singleton
+    {
+    public:
+        // Non-copyable, non-movable
+        singleton(singleton const&) = delete;
+        singleton(singleton&&) = delete;
+        singleton& operator=(singleton const&) = delete;
+        singleton& operator=(singleton&&) = delete;
+
+        static std::shared_ptr<T> instance()
```

Comment:
It might be better to return a `const unique_ptr<T>&` here.
Reference counting will delay access in SMP scenarios and shared ownership isn't strictly necessary here.
Especially when we're obfuscating it with a macro.

Reply:
I've added a CRTP based solution. Now the `CvsExtern` can be directly replaced with `CMK_GENERATE_SINGLETON` and `CvsAccess` with `CMK_ACCESS_SINGLETON` to achieve the exact same solution as the previous macros did (with Singleton behavior).